Health api memory optimisation

SwethaGuptha opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
 
https://github.com/opensearch-project/OpenSearch/issues/11684 - on a large cluster the memory allocation of TransportClusterHealthAction is around 33%. Memory/compute requirement of TransportClusterHealthAction is proportional to the number of indices and shards on the cluster. Transport ClusterHealthResponse maintains health objects for individual indices and shards that can increase the memory allocation of this with the growth of indices and shards count.  At rest layer, there is a room for optimization for _cluster/health API when request is at cluster/indices level by performing inline aggregations and not creating and maintaing shard health objects. 

Changes:
- Introducing a flag ```honourClusterHealthLevel``` in ClusterHealthRequest to identify if the request param ```level```  should be honored or not when computing health response in TransportClusterHealthAction - Currently ```level``` param is only used at REST layer for generating response from the transport response based on the level requested ([usage](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/admin/cluster/health/ClusterHealthResponse.java#L455)) and has no meaning at transport layer. This flag will help TransportClusterHealthAction in differentiating if indices/shards level health information is required in the transport response or not. By default this flag is set to false to keep the change backward compatible.
- Initializing ClusterIndexHealth without creating the ClusterShardHealth objects per shard when requested level for cluster health is not at shards level, health aggregation are made inline.
-  ClusterHealthResponse computes the delayed unassigned shards information by iterating over all shards of the cluster([reference](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/admin/cluster/health/TransportClusterHealthAction.java#L533)).This is avoided by computing the delayed unassigned shards along with computation of other shard states in ClusterStateHealth. 
- Using index based iteration for iterating over ShardRoutings of a shard. For each loop iteration uses an iterator under the hood and since the ShardRouting list maintained in the IndexShardRoutingTable is un-modifiable, a new iterator object is created on the heap.

Memory allocation 5 min profile of an idle cluster before and after optimisation
**Before Optimisation**
<img width="1918" alt="Base Profile" src="https://github.com/user-attachments/assets/fb7612dc-b867-482b-8a56-b2fe09820ce2">
**After Optimisation**
<img width="959" alt="Memory Optimisation Profile" src="https://github.com/user-attachments/assets/559b6dea-f11d-449e-88c5-bd394c51758b">

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
<!-- List any other related issues here -->

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

gbbafna commented:
> Introducing a flag honourClusterHealthLevel in ClusterHealthRequest to identify if the request param level should be honored or not when computing health response in TransportClusterHealthAction - Currently level param is only used at REST layer for generating response from the transport response based on the level requested ([usage](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/admin/cluster/health/ClusterHealthResponse.java#L455)) and has no meaning at transport layer. This flag will help TransportClusterHealthAction in differentiating if indices/shards level health information is required in the transport response or not. By default this flag is set to false to keep the change backward compatible.

How are breaking backward compatibility when the flag is true ? Isn't this construct only used in `Transport` and not in `Rest` ?

gbbafna commented on server/src/internalClusterTest/java/org/opensearch/cluster/ClusterHealthIT.java:
```diff
@@ -439,4 +442,130 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         }
     }
+
+    public void testHealthAtLevelParamEqualsClusterIsHonoured() {
```
Lets add `AWARENESS_ATTRIBUTES` level IT as well .

sumitasr replied:
+1, since there are 4 levels.

gbbafna commented on server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java:
```diff
@@ -129,6 +135,80 @@ public ClusterStateHealth(final ClusterState clusterState, final String[] concre
+        numberOfDataNodes = clusterState.nodes().getDataNodes().size();
+        hasDiscoveredClusterManager = clusterState.nodes().getClusterManagerNodeId() != null;
+        indices = new HashMap<>();
+        boolean isIndexLevelHealthInfo = healthLevel == ClusterHealthRequest.Level.INDICES
```
should we rename to `isIndexOrShardLevelHealthInfo`

sumitasr reviewed (changes requested):
Unit tests are missing for all the changes in source code. Can we please cover those for files where we have added custom logic to iterate and compute health?

UTs covering corner cases will help i guess if you can check on coverage once.

SwethaGuptha commented on server/src/main/java/org/opensearch/cluster/health/ClusterIndexHealth.java:
```diff
@@ -189,6 +189,102 @@ public ClusterIndexHealth(final IndexMetadata indexMetadata, final IndexRoutingT
+            }
+        }
+
+        if (indexRoutingTable.shards().isEmpty()) { // might be since none has been created yet (two phase index creation)
```
IndexRoutingTable constructors would ensure shards map is always initialized. Added the check to be safe.

shwetathareja commented on server/src/main/java/org/opensearch/action/admin/cluster/health/ClusterHealthRequest.java:
```diff
@@ -76,6 +76,17 @@ public class ClusterHealthRequest extends ClusterManagerNodeReadRequest<ClusterH
+     *                  For Level.SHARDS, information on indices and shards will be returned to the transport client
+     * By default, the flag is disabled.
+     */
+    private boolean honourClusterHealthLevel = false;
```
why do you need the flag? The levels are already present, i thought when level is cluster, shard level information is not needed at all?

shwetathareja commented on server/src/main/java/org/opensearch/cluster/health/ClusterIndexHealth.java:
```diff
@@ -189,6 +189,102 @@ public ClusterIndexHealth(final IndexMetadata indexMetadata, final IndexRoutingT
+            for (IndexShardRoutingTable indexShardRoutingTable : indexRoutingTable) {
+                int activeShardsPerShardId = 0;
+
+                List<ShardRouting> shardRoutings = indexShardRoutingTable.shards();
+                int shardRoutingCountPerShardId = shardRoutings.size();
+                for (int index = 0; index < shardRoutingCountPerShardId; index++) {
+                    ShardRouting shardRouting = shardRoutings.get(index);
+                    if (shardRouting.active()) {
+                        computeActiveShards++;
+                        activeShardsPerShardId++;
+                    } else if (shardRouting.relocating()) {
+                        computeRelocatingShards++;
+                    } else if (shardRouting.initializing()) {
+                        computeInitializingShards++;
+                    } else if (shardRouting.unassigned()) {
+                        computeUnassignedShards++;
+                        if (shardRouting.unassignedInfo().isDelayed()) {
+                            computeDelayedUnassignedShards++;
+                        }
+                    }
+                }
```
all this code will be duplicated across ClusterIndexHealth & ClusterShardHealth?

shwetathareja replied:
We should think about keeping the information calculated as part of IndexRoutingTable using IndexShardRoutingTable itself. It can serve as cache as well. We should skip this intermediate step to just optimize ClusterStateHealth object instantiation

SwethaGuptha replied:
This duplication is arising from how the data models are defined and couldn't have been avoided without introducing some other abstract classes like cache, this requires holistic thinking to weigh its benefits as it will have its own memory foot-print and comes with challenge's of keeping the cache upto date. 

Given the performance benefits we have observed with the in-line aggregations, do you think caching change can be done incrementally?

SwethaGuptha replied:
Git hub issue created for exploring cache based solution: https://github.com/opensearch-project/OpenSearch/issues/15685

shwetathareja commented on server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java:
```diff
@@ -129,6 +135,80 @@ public ClusterStateHealth(final ClusterState clusterState, final String[] concre
+            if (indexRoutingTable == null) {
+                continue;
+            }
+
+            ClusterIndexHealth indexHealth = new ClusterIndexHealth(indexMetadata, indexRoutingTable, healthLevel);
+            computeActivePrimaryShards += indexHealth.getActivePrimaryShards();
+            computeActiveShards += indexHealth.getActiveShards();
+            computeRelocatingShards += indexHealth.getRelocatingShards();
+            computeInitializingShards += indexHealth.getInitializingShards();
+            computeUnassignedShards += indexHealth.getUnassignedShards();
+            computeDelayedUnassignedShards += indexHealth.getDelayedUnassignedShards();
+            if (indexHealth.getStatus() == ClusterHealthStatus.RED) {
+                computeStatus = ClusterHealthStatus.RED;
+            } else if (indexHealth.getStatus() == ClusterHealthStatus.YELLOW && computeStatus != ClusterHealthStatus.RED) {
+                computeStatus = ClusterHealthStatus.YELLOW;
+            }
```
All this seems to be duplicate code across constructors

SwethaGuptha replied:
Same as https://github.com/opensearch-project/OpenSearch/pull/15492#discussion_r1744120491

shwetathareja commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterHealthAction.java:
```diff
@@ -105,6 +105,7 @@ public static ClusterHealthRequest fromRequest(final RestRequest request) {
         if (level != null) {
             clusterHealthRequest.setLevel(level);
         }
+        clusterHealthRequest.setHonourClusterHealthLevel(true);
```
applyLevelAtTranportLayer

SwethaGuptha replied:
Addressed.

SwethaGuptha commented:
> Unit tests are missing for all the changes in source code. Can we please cover those for files where we have added custom logic to iterate and compute health?
> 
> UTs covering corner cases will help i guess if you can check on coverage once.

Added UTs.

SwethaGuptha commented on server/src/main/java/org/opensearch/cluster/health/ClusterIndexHealth.java:
```diff
@@ -189,6 +189,104 @@ public ClusterIndexHealth(final IndexMetadata indexMetadata, final IndexRoutingT
+    }
+
+    public static ClusterHealthStatus getIndexHealthStatus(ClusterHealthStatus shardHealth, ClusterHealthStatus computeStatus) {
+        if (shardHealth == ClusterHealthStatus.RED) {
+            return ClusterHealthStatus.RED;
+        } else if (shardHealth == ClusterHealthStatus.YELLOW && computeStatus != ClusterHealthStatus.RED) {
+            // do not override an existing red
+            return ClusterHealthStatus.YELLOW;
+        }
+        return computeStatus;
```
Addressed.

sumitasr commented on server/src/internalClusterTest/java/org/opensearch/cluster/ClusterHealthIT.java:
```diff
@@ -439,4 +442,154 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         }
     }
+
+    public void testHealthAtLevelParamEqualsClusterIsHonoured() {
```
Minor - Should we change IT names to reflect the update in the param names in latest changes. So, it comes becomes clear when read in future.

E.g. testHealthWithClusterLevelAppliedAtTransportLayer?

Applicable in all the IT names.

SwethaGuptha replied:
Addressed.

sumitasr reviewed (approved):
Thanks @SwethaGuptha for making the changes.

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/health/ClusterHealthResponse.java:
```diff
@@ -237,6 +237,26 @@ public ClusterHealthResponse(
+    public ClusterHealthResponse(
+        String clusterName,
+        String[] concreteIndices,
+        ClusterHealthRequest clusterHealthRequest,
```
Would prefer to have this abstracted in ClusterHealthResponse  itself to avoid branching at the caller action.

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/health/ClusterHealthRequest.java:
```diff
@@ -76,6 +76,17 @@ public class ClusterHealthRequest extends ClusterManagerNodeReadRequest<ClusterH
+     * This flag will be used by the TransportClusterHealthAction to decide if indices/shards info is required in the ClusterHealthResponse or not.
+     * When the flag is disabled - indices/shard info will be returned in ClusterHealthResponse regardless of the health level requested.
+     * When the flag is enabled - indices/shards info will be set according to health level requested.
+     *                  For Level.CLUSTER (or) LevelAWARENESS_ATTRIBUTES - information on indices/shards will NOT be returned to the transport client
```
Addressed.

shwetathareja commented:
@SwethaGuptha Add Changelog for it.

SwethaGuptha commented:
> ❌ Gradle check result for [9ad4739](https://github.com/opensearch-project/OpenSearch/commit/9ad47394fcac8b98538302bbeca91b4a1c9c1d4e): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/47271/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Failed test case:
[org.opensearch.cluster.ClusterHealthIT.testHealthWithShardLevelAppliedAtTransportLayer](https://build.ci.opensearch.org/job/gradle-check/47271/testReport/junit/org.opensearch.cluster/ClusterHealthIT/testHealthWithShardLevelAppliedAtTransportLayer/)
[org.opensearch.cluster.ClusterHealthIT.testHealthWithShardLevelAppliedAtTransportLayer](https://build.ci.opensearch.org/job/gradle-check/47271/testReport/junit/org.opensearch.cluster/ClusterHealthIT/testHealthWithShardLevelAppliedAtTransportLayer_2/)
[org.opensearch.cluster.ClusterHealthIT.testHealthWithShardLevelAppliedAtTransportLayer](https://build.ci.opensearch.org/job/gradle-check/47271/testReport/junit/org.opensearch.cluster/ClusterHealthIT/testHealthWithShardLevelAppliedAtTransportLayer_3/)
[org.opensearch.cluster.ClusterHealthIT.testHealthWithShardLevelAppliedAtTransportLayer](https://build.ci.opensearch.org/job/gradle-check/47271/testReport/junit/org.opensearch.cluster/ClusterHealthIT/testHealthWithShardLevelAppliedAtTransportLayer_4/)
[org.opensearch.index.shard.RemoteStoreRefreshListenerTests.testRefreshSuccessAfterFailureInFirstAttemptAfterSnapshotAndMetadataUpload](https://build.ci.opensearch.org/job/gradle-check/47271/testReport/junit/org.opensearch.index.shard/RemoteStoreRefreshListenerTests/testRefreshSuccessAfterFailureInFirstAttemptAfterSnapshotAndMetadataUpload/)
[org.opensearch.index.shard.RemoteStoreRefreshListenerTests.classMethod](https://build.ci.opensearch.org/job/gradle-check/47271/testReport/junit/org.opensearch.index.shard/RemoteStoreRefreshListenerTests/classMethod/)
[org.opensearch.index.shard.RemoteStoreRefreshListenerTests.classMethod](https://build.ci.opensearch.org/job/gradle-check/47271/testReport/junit/org.opensearch.index.shard/RemoteStoreRefreshListenerTests/classMethod_2/)

SwethaGuptha commented:
> ❌ Gradle check result for [5f5f4f8](https://github.com/opensearch-project/OpenSearch/commit/5f5f4f872ebe27d6f6525c3c4e5b4e7d5d316eb0): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/47332/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test https://github.com/opensearch-project/OpenSearch/issues/15731

SwethaGuptha commented:
x

> ❌ Gradle check result for [0bf50eb](https://github.com/opensearch-project/OpenSearch/commit/0bf50eb6d5c096e325914e72e99199205e9f5564): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/47316/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flajy test https://github.com/opensearch-project/OpenSearch/issues/15757

SwethaGuptha commented:
> ❕ Gradle check result for [51db30c](https://github.com/opensearch-project/OpenSearch/commit/51db30c7e276248cb647492f9c0c822369bf7bde): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/47396/)
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

Flaky test https://github.com/opensearch-project/OpenSearch/issues/14306

SwethaGuptha commented:
> ❌ Gradle check result for [75caa4a](https://github.com/opensearch-project/OpenSearch/commit/75caa4a218f4c38bd0934364c90ebf7e5283276a): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/47430/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test https://github.com/opensearch-project/OpenSearch/pull/15795
